### ISSUE TITLE
Give packages means to alter nsswitch.conf

### ISSUE THREAD

**Issue by agoode** [pbrezina/authselect#51]
I learned about authselect from this bug:
https://bugzilla.redhat.com/show_bug.cgi?id=1577243

I would like to have nss-mdns stop editing nsswitch.conf directly in Fedora. Does authselect provide a mechanism to allow for this? Could I make nss-mdns depend on authselect?

**Comment by pbrezina** [comment]
Thank you for coming to us. We would like to own nsswitch.conf in future Fedora so it would be good to solve this kind of things.

There is a possibility of modifying existing profiles in several ways (see [authselect-profiles(5)](https://github.com/pbrezina/authselect/blob/master/src/man/authselect-profiles.5.txt.in.in) but maybe, it would be best to add it to upstream profiles. What is the purpose of this module?

**Comment by pbrezina** [comment]
I was thinking about this issue. Shipping a profile in nss-mdns is possible, but not a good option, since you only modify the host map, so you would have to ship custom profiles for both sssd, winbind and any future upstream profile. And it would not be wise if such module meddled with currently selected profile.

Good solution, for you anyway, would be to include this directly in upstream profiles on condition, something like this:
```
hosts: files {if "with-mdns":mdns4_minimal [NOTFOUND=return]} dns {if "with-mdns":mdns4} myhostname
```

But this is not very good solution for upstream, since we would have to manage all possible requested modules this way which we simply do not want to. As I expect there will be such requests, we have to come up with a better solution, give me some time and maybe we will find out something that suites both sides.

However, in my opinion, rpm packages ideally should not meddle with nsswitch (as you wrote, it is quite brittle) since you can't ensure that the change works correctly with every possible combination of enabled modules. With authselect, users are expected to install a module and if not supported by upstream profile, create a custom profile with correct setting. Would you consider this way?

**Comment by orbisvicis** [comment]
I think upstream profile is [this repository](https://github.com/pbrezina/authselect/tree/master/profiles) but the fedora package maintains [several patches as well](https://src.fedoraproject.org/cgit/rpms/authselect.git/tree/). Expanding on @pbrezina's comment:

* For each upstream profile (winbind,sssd) add

      hosts: files {if "with-mdns":mdns4_minimal [NOTFOUND=return]} dns {if "with-mdns":mdns4} myhostname

  * `/usr/share/authselect/default/sssd/nsswitch.conf`
  * `/usr/share/authselect/default/winbind/nsswitch.conf`
* Also update each README
* Then for nss-mdns scripts:

      %post
      /sbin/ldconfig
      # It is possible to activate non-existent features but not double-activate
      # features. So verify for supported profile.
      case "$(authselect current | sed -n 's/^Profile ID:\s*\(.*\)$/\1/p')" in
          sssd|winbind)
              # If already activated, this won't do anything.
              authconfig enable-feature with-mdns
              # TODO: I think this will create a new 'nsswitch.conf'; double-check.
              ;;
      esac
  
      %preun
      case "$(authselect current | sed -n 's/^Profile ID:\s*\(.*\)$/\1/p')" in
          sssd|winbind)
              # If already activated (i.e. before nss-mdns installation) this will
              # nonetheless disable it.
              authconfig disable-feature with-mdns
              # TODO: I think this will create a new 'nsswitch.conf'; double-check.
              ;;
      esac

* Ensure nss-mdns depends on authselect
* Perhaps give authselect a weak dependency on nss-mdns

**Comment by pbrezina** [comment]
Sorry for late update, I was busy with other tasks, I will focus on authselect now for next few weeks. The basic idea is to provide a way to include some publicly available file(s) into authselect nsswitch.conf. I yet need to come with a concrete design that I will then discuss with you if it is suitable for package maintainers.

**Comment by akors** [comment]
> I yet need to come with a concrete design that I will then discuss with you if it is suitable for package maintainers.

Please also consider the users: my use case is changing the order of only the hosts line in nsswitch.conf, but I would have to create and maintain a whole custom profile just for that line. I "don't care" about most of nsswitch.conf (or rather, I trust the packagers choices).

I don't think the monolithic "profiles" approach is a very good one, when packages and users often only require certain small changes instead of a whole profile.

> The basic idea is to provide a way to include some publicly available file(s) into authselect nsswitch.conf.

What about the classic `.d` config directory approach, for example by dropping override files into `/etc/nsswitch.conf.d`?

**Comment by pbrezina** [comment]
> Please also consider the users: my use case is changing the order of only the hosts line in nsswitch.conf, but I would have to create and maintain a whole custom profile just for that line. I "don't care" about most of nsswitch.conf (or rather, I trust the packagers choices).

Maintaining a profile is not that hard, as you can symlink pam configuration and only maintain custom nsswitch.conf. But this ticket will remove the need to create custom profile for nsswitch.conf modifications.

> I don't think the monolithic "profiles" approach is a very good one, when packages and users often only require certain small changes instead of a whole profile.

Systems are usually configured only once, so you can even use authselect once to boostrap your configuration and then do changes in files. It will not be overwritten by authselect unless you specify `--force` option on command line. On the other side, enterprise environments should benefit from profiles as they will gain better control over the resulting configuration.

**Comment by pbrezina** [comment]
I prepared a design documents. Any thoughts are welcomed.
https://github.com/pbrezina/authselect/wiki/Design-Document:-nsswitch.conf-modification

I will be on PTO next week so my response will be slow.

**Comment by agoode** [comment]
Thank you, I will take a look.

**Comment by agoode** [comment]
I read the doc, thanks for writing it. Modifying nsswitch.conf does seem like a hard problem, especially given the ordering issues you describe.

Could there be a way of resolving these conflicts with a constraint solver? A module could express Before and After constraints possibly. I don't really know.

Are there changes I could make to nss-mdns to make it more authselect compatible? I think adding it to the upstream profile is looking more and more like the solution.

**Comment by akors** [comment]
> I prepared a design documents. Any thoughts are welcomed. Design Document: nsswitch.conf modification

Maybe I didn't quite understand it right, but if you allow all programs and even human admins to write to /etc/authselect/nsswitch-changes, don't you just move the problem of conflicts somewhere else? How do packages ensure that they're not screwing up admin configuration and vice-versa?

Personally I believe that multiple packages+humans writing to the same file is not good design. I think systemd got it right with their drop-in configuration., one (ore more) file per package

> When an authselect profile is selected, it will merge profile template with this file and generate nsswitch.conf

What does this merging look like?

> If a profile that overwrites some nsswitch map is selected, the user value is still present in the file, it is just not applied and not present in the generated /etc/nsswitch.conf

So you are basically shadowing peoples changes? Isn't that a bit confusing? I can imagine scenarios where people change nsswitch.conf (like they used to in the old days), but their changes have no effect due to this shadowing.

**Comment by pbrezina** [comment]
I'm back from PTO. Thank you for your  comments.

> Maybe I didn't quite understand it right, but if you allow all programs and even human admins to write to /etc/authselect/nsswitch-changes, don't you just move the problem of conflicts somewhere else? How do packages ensure that they're not screwing up admin configuration and vice-versa?

Writing to this file is the same as current situation, i.e. both admins and packages are writing to nsswitch.conf.  The one difference is that it can be overwritten by profile. For example, if lets have a custom profile  called `files-first` that wants to call `files` before `sss`:

```
$ cat /etc/authselect/custom/files-first/nsswitch.conf
passwd: files sss
group: files sss

$ cat /etc/authselect/nsswitch-changes
passwd: sss files
group: sss files
hosts: files dns myhostname

$ authselect select custom/files-first
$ cat /etc/nsswitch.conf
passwd: files sss
group: files sss
hosts: files dns myhostname
```

Now mdns module is installed, it writes to nsswitch-changes and calls `authselect apply-changes`

```
$ cat /etc/authselect/nsswitch-changes
passwd: sss files
group: sss files
hosts: files mdns4_minimal [NOTFOUND=return] dns mdns4

$ authselect apply-changes
$ cat /etc/nsswitch.conf
passwd: files sss
group: files sss
hosts: files mdns4_minimal [NOTFOUND=return] dns mdns4
```
So the profile configuration is the one that always takes precedence. If admistrator wants to ensure that his configuration is not touched, he need to create new profile (there is https://github.com/pbrezina/authselect/pull/62 that implements nice cli tool to help with this step).

> Personally I believe that multiple packages+humans writing to the same file is not good design. 

Yes, I agree with you here. Therefore humans should create custom profile, if they want to have full control over the configuration.

> akora: I think systemd got it right with their drop-in configuration., one (ore more) file per package
> agoode: Could there be a way of resolving these conflicts with a constraint solver? A module could express Before and After constraints possibly. I don't really know.

This is nice idea and I agree that systemd unit files are good. But this is something very problematic for nsswitch.conf because the use case is different. In systemd, there are services that are running simultaneously, just some service A depends on service B so it needs to be run After service B. This is not solving conflicts, but dependencies. In nsswitch.conf we do not have dependencies, we have conflicts. Both `files sss`  and `sss files` is a valid configuration as is `files mdns4_minimal [NOTFOUND=return] dns mdns4` and `mdns4_minimal [NOTFOUND=return] mdns4 dns files` but  which of these two administrator really wants to have? This cannot be solved. There is default and recommended configuration, but the order is really arbitrary.

However, even though it can not be used for managing nsswitch.conf all the time, I think something like this may be doable for package installation. Package would specify where the module should be placed and authselect would change nsswitch-changes for it so the package would not need to parse and modify the file. (Not in scope of this release but maybe for Fedora 30).



>> If a profile that overwrites some nsswitch map is selected, the user value is still present in the file, it is just not applied and not present in the generated /etc/nsswitch.conf
>
> So you are basically shadowing peoples changes? Isn't that a bit confusing? I can imagine scenarios where people change nsswitch.conf (like they used to in the old days), but their changes have no effect due to this shadowing.

If they modify `/etc/nsswitch.conf` authselect will never overwrite the changes without user involvement as it requires `--force` parameter to do so. If they modify `/etc/authselect/nsswitch-changes` they will be aware of this because they are willingly using authselect.

There is also one thing to be kept in mind. Single-machine users do not need to use authselect and it makes perfect sense not to use it. The machine is configured with authselect by anaconda to provide default setup but users can modify nsswitch.conf and PAM as they wish. Both nsswitch.conf and PAM is something that is configured once and then it is usually left intact for the machine life-time. So authselect use case for standard Fedora users is to provide an initial configuration and it can be then modified as desired. Profile-based configuration makes sense in corporate sphere when you want to ensure that exact configuration is present on all machines.

**Comment by pbrezina** [comment]
FYI I also opened https://github.com/pbrezina/authselect/issues/77 after pushing the patches. I will prepare a change page for Fedora 30 which would give packages some tooling for modifying nsswitch.conf, so package maintainers will not have to worry about authselect/nsswitch.conf at all.
